### PR TITLE
#170947674 Send notification for the approved request

### PR DESCRIPTION
**Pull request by T2Wil**
#### What does this PR do?
- This PR enables a manager to approve the request and then notify the requester via the email and in-app.
#### Description of Task to be completed?
- Create an endpoint for a manager to approve the request
ENDPOINT: ```PATCH http://localhost:4000/api/v1/trips/approve/:requestId```.
- Create an endpoint for a requester to view a specific request.
ENDPOINT: ```GET http://localhost:4000/api/v1/trips/request/:requestId```.
- Send an email notification to the requester about the approved request.
- Notify the requester instantly in the app.
- Test the two endpoints.
- Document the two endpoints.

#### How should this be manually tested?
- open your terminal and run this command `git clone https://github.com/andela/knightsbn    backend.git` to clone repository.
- run `cd knights-bn-backend` to get inside the project.
- run `git fetch && git checkout ft-notification-for-approved-request-170947674` to get inside this branch
- run `npm install` to install all dependencies
- check if you have PostgresQL database if you don't have it install it from here [PostgreSQL](https://www.2ndquadrant.com/en/blog/pginstaller-install-postgresql/)
- create2  Postgres databases locally: one for development and another for testing .
- add `BN_EMAIL_NO_REPLY=no-reply@brftnomad.com` in your .env or create a .env file, paste in below variables and replace xxxx with the correct values: (Check code-knights `slack channel` to find the correct values)

```
DB_USER='name of the user of the database'
DB_PASS='password of your Postgres'
DB_NAME='name of the database || Barefoot'
DB_TEST_NAME='name of the database for test || Barefoot_test'
NODE_ENV='development'
SECRETKEY='PrivateKey'
GOOGLE_CLIENT_ID = 'XXXXX'
GOOGLE_CLIENT_SECRET = 'XXXXX'
GOOGLE_CALLBACK_URL = 'XXXXX'
FCBK_CLIENT_ID = 'XXXXX'
FCBK_CLIENT_SECRET = 'XXXXX'
FCBK_CALLBACK_URL = 'XXXXX'
BN_API_KEY= 'XXXX'
MANAGER_PSWD= 'XXXX'
```
- run `npm run createTables` to create the database tables and seed in primary data.
- run `npm dev-start` to start the server.

1. Log in as a requester to create a new trip
POST : ` http://localhost:4000/api/v1/auth/login`
```
{
    "email": "ishimwewil005@gmail.com",
    "password": "Password@1"
}
```
image

2. Copy the token and paste it in the below Codepen code to track in-app notification.
[Codepen link](https://codepen.io/king_Bomaye/pen/VwLbQGX?editors=0011)

Replace the token value with your token in the Codepen.
image

3. Create a new trip
POST `http://localhost:4000/api/v1/trips/returnTrip`
```
{
	"origin": "Kigali",
	"destination": "Kampala",
	"departureDate": "2020-03-10",
	"returnDate": "2020-03-29",
	"reason": "Having fun",
	"accommodation": "A Hotel"
}
```
image

4. log in as an admin
POST : ` http://localhost:4000/api/v1/auth/login`

```
{
    "email": "william.ishimwe@andela.com",
    "password": "Password@1"
}
```
image

5. View pending requests and pick id of one of the pending request.
GET : ` http://localhost:4000/api/v1/trips/pendingApproval`
image

6.  Approve request.
From the above pending request, let's approve the request with the id of 5. N:B: make sure is the one we pasted his token in the [Codepen link](https://codepen.io/king_Bomaye/pen/VwLbQGX?editors=0011).
image

7. Received email:
image

8. In-app notification:
image

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?

[#170947674](https://www.pivotaltracker.com/story/show/170947674)

### CODE REVIEW

**Review by Kagorora — APPROVED** [review]
@T2Wil  You did a good job by creating helpers for often used function and your work can be merged

**Comment by EugeneMunya** [comment]
Screen Shot 2020-03-11 at 12 19 11
make it clickable as story criterial says below
Screen Shot 2020-03-11 at 12 21 06

**Comment by EugeneMunya** [comment]
Screen Shot 2020-03-11 at 12 28 07
@T2Wil  how will you get that token you copy and past in HTML file?

**Comment by EugeneMunya** [comment]
Screen Shot 2020-03-11 at 12 29 26
Make it easy for reviewers to copy that URL

**Comment by T2Wil** [comment]
> Screen Shot 2020-03-11 at 12 28 07
> 
> @T2Wil how will you get that token you copy and past in HTML file?

Hey @EugeneMunya , that token is the one you get when you log in as a requester; in the above case was the one you received with the given user from PR description:
```
{
    "email": "ishimwewil005@gmail.com",
    "password": "Password@1"
}
```

**Comment by T2Wil** [comment]
> Screen Shot 2020-03-11 at 12 29 26
> 
> Make it easy for reviewers to copy that URL

Hey @EugeneMunya , the link was provided on step 2, you could have seen it but for ease It is now on the step 6 too. Thanks.

**Comment by T2Wil** [comment]
> Screen Shot 2020-03-11 at 12 19 11
> 
> make it clickable as story criterial says below
> Screen Shot 2020-03-11 at 12 21 06

Thanks @EugeneMunya , I missed it. I will let you know when I am done!

**Comment by EugeneMunya** [comment]
> > Screen Shot 2020-03-11 at 12 28 07
> > @T2Wil how will you get that token you copy and past in HTML file?
> 
> Hey @EugeneMunya , that token is the one you get when you log in as a requester; in the above case was the one you received with the given user from PR description:
> 
> ```
> {
>     "email": "ishimwewil005@gmail.com",
>     "password": "Password@1"
> }
> ```
I am asking how you will insert that token in HTM in real App ? for now we have to copy and past it in html

**Comment by T2Wil** [comment]
> > Screen Shot 2020-03-11 at 12 19 11
> > make it clickable as story criterial says below
> > Screen Shot 2020-03-11 at 12 21 06
> 
> Thanks @EugeneMunya , I missed it. I will let you know when I am done!

Please @EugeneMunya , check the again; the issue is solved!

File: .eslintrc.js [hunk 1]
```diff
@@ -1,6 +1,12 @@
...
+  ecmaFeatures: {
+    modules: true,
+    spread: true,
+    restParams: true
+  },
+  root: true,
+  'extends': 'airbnb-base',
   env: {
-    es6: true,
     node: true,
```
Comment (kagabof) [inline .eslintrc.js:11]: Why did you remove es6?

File: src/db/seeders/20200225213741-two_way_trip.js [hunk 2]
```diff
@@ -8,19 +8,20 @@ module.exports = {
     gender: 'male',
     email: 'william.ishimwe@andela.com',
     passport: 'null',
-    password: process.env.MANAGER_PSWD,
+    password: '$2b$10$SO7hwwylEnHKZIv68w/ePOK4yybPnhXa5.tQSUBe.PSb6FBKyuVO.',
```
Comment (kagabof) [inline src/db/seeders/20200225213741-two_way_trip.js:11]: Why did you remove env?

File: src/db/seeders/20200225213741-two_way_trip.js [hunk 3]
```diff
@@ -8,19 +8,20 @@ module.exports = {
...
-    lastName: 'Eric',
+    firstName: 'William',
+    lastName: 'Ishimwe',
     role: 'requester',
     gender: 'male',
-    email: 'ishimwe@gmail.com',
+    email: 'ishimwewil005@gmail.com',
     passport: '1940567',
-    password: process.env.MANAGER_PSWD,
+    password: '$2b$10$SO7hwwylEnHKZIv68w/ePOK4yybPnhXa5.tQSUBe.PSb6FBKyuVO.',
```
Comment (kagabof) [inline src/db/seeders/20200225213741-two_way_trip.js:23]: Why did you remove env?

File: src/helpers/errorHandler.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  if (error.message.includes('invalid input syntax for type integer')) {
+    return res.status(422).json({
+      status: res.statusCode,
+      error: 'make sure the request id entered is a number.'
+    });
+  }
+  return res.status(500).json({ error });
+};
+export default handleError;
```
[the remark this answers is not in the document]
Comment (kagabof) [reply src/helpers/errorHandler.js]: please resolve this.

File: src/routes/trips.js [hunk 5]
```diff
@@ -531,32 +615,28 @@ import commentValidate from '../middlewares/newComment';
...
 import {
-  validateRequestDate, validateCityDate, tripInformation, multicity, checkIfRequestExists,
+  validateRequestDate, validateCityDate, tripInformation, multicity, checkIfRequestExists
 } from '../middlewares/validateDate';
-
+import validateParams from '../middlewares/validateParams';
 
 const router = express.Router();
+const { filterTrips } = requestsController;
+const { getRequestStatus, markAsRead } = requestStatusController;
```
Comment (kagabof) [inline src/routes/trips.js]: Please do this while importing it.

**Review by kagabof — CHANGES_REQUESTED** [review]

File: src/db/seeders/20200225213741-two_way_trip.js (hunk 2 above)
Comment (T2Wil) [reply src/db/seeders/20200225213741-two_way_trip.js:11]: There was trouble verifying that password fetched from .env and pasting that hashed password directly into testing seeds solved the issue.

File: src/db/seeders/20200225213741-two_way_trip.js (hunk 3 above)
Comment (T2Wil) [reply src/db/seeders/20200225213741-two_way_trip.js:23]: There was trouble verifying that password fetched from .env and pasting that hashed password directly into testing seeds solved the issue.

File: .eslintrc.js (hunk 1 above)
Comment (T2Wil) [reply .eslintrc.js:11]: By mistake but fixed!

File: src/routes/trips.js (hunk 5 above)
Comment (T2Wil) [reply src/routes/trips.js]: Fixed!

**Review by kagabof — APPROVED** [review]
